docs/development/design/build-images.rst

```diff
@@ -191,94 +299,122 @@ Examples of these versions are:
+We may want to keep only the latest Ubuntu LTS releases available in production,
+with a special consideration for our current Ubuntu 18.04 LTS on ``stable``, ``latest`` and ``testing`` because 100% of the projects depend on them currently.
+Once Ubuntu 22.04 LTS is released, we should communicate that Ubuntu 20.04 LTS is deprecated,
+and give users 2 years to migrate to a newer image (1 year less than the official Ubuntu support for LTS images)
 
-We could keep ``stable``, ``latest`` and ``testing`` for some time without worry too much.
-New projects shouldn't be able to select these images and they will be forced to use ``ubuntu20``
-if they don't specify one.
+.. warning::
 
-We may want to keep the two latest Ubuntu LTS releases available in production.
-At the moment of writing this they are:
+   Deleting ``-base`` images from the build servers **will make project's builds to fail**.
+   We want to keep supporting them as much as we can, but having a well-defined deprecation policy is a win.
 
-* Ubuntu 18.04 LTS (our ``stable``, ``latest`` and ``testing`` images)
-* Ubuntu 20.04 LTS (our new ``ubuntu20``)
 
-Once Ubuntu 22.04 LTS is released, we should deprecate Ubuntu 18.04 LTS,
-and give users 6 months to migrate to a newer image.
+Work required and rollout plan
+------------------------------
 
+The following steps are required to support the full proposal of this document.
 
-Work required
--------------
+#. allow users to install extras languages requirements via config file
 
-There are a lot of work to do here.
-However, we want to prioritize it based on users' impact.
+   * update config file to support ``build.os`` and ``build.languages`` config
+   * modify builder code to run ``asdf install`` for all supported languages
 
-#. allow users to install packages with APT
+#. build a new base Docker image with new structure (``ubuntu20-base``)
 
-   * update config file to support ``build.system_packages`` config
-   * modify builder code to run ``apt-get install`` as ``root`` user
+   * build new image with Ubuntu 20.04 LTS and pre-installed ``asdf`` with all its plugins
+   * do not install any language version on base image
+   * deploy builders with new base image
 
-#. allow users to install extras via config file
+#. update builders to install ``build.languages`` selected by the user
 
-   * update config file to support ``build.extras`` config
-   * modify builder code to run ``nodenv install`` / ``rustup install``
-   * re-build our current images with pre-installed nodenv and rustup
-   * make sure that all the versions are the same we have in production
-   * deploy builders with newer images
 
-#. pre-build commands (not covered in this document)
+At this point, we will have a full working setup.
+It will be opt-in by using the new Config File V3.
+However, *all languages* will be installed at build time;
+which will "penalize" all projects because all of them will have to install Python.
```

Comment:
I don't think we can release this with a 2 minute penalty for installing Python unless this is explicitly defined as a testing or beta phase. We definitely need to either pre-compile those docker images, or have a cache for Python I think. Otherwise I doubt any users will use it willingly.

Reply:
Yes. This is for testing to know that new images are working, `asdf` is installing all the languages, etc.

At this point we can add some projects that we already know they require `nodejs` and do some QA with them before moving forward to the next phase.